### DISCUSSION TITLE
Testing and demonstrating immutability of blockfile

### DISCUSSION THREAD

**Discussion by manavdeep** [hyperledger/fabric#5051] [Q&A] [answered]
So anyone tried to demonstrate the immutability aspect of hyperledger to their team?
Like there are 2 ways going about it:
1. Tampering the ledger on a compromised peer
2. Tampering the world state on a compromised peer

For the moment lets talk about 1st. I was hoping the following should happen:
1. Enter a compromised peer and manually edit and save it blockfile in container.
2. Next transaction should fail (ofcourse the endorsement policy is depending on the tampered peer's participation)

Its not happening unless I restart the tampered peer. Things keep working fine. The blockfile that I tampered shows the time stamp of old data only. So whats happening? Is the peer operating out of memory and not writing to the disk?

**Comment by denyeart** [comment, marked as answer]
The peer does not check old blocks at runtime, it just verifies blocks as they are processed and validates transactions relative to its state database.
To verify hashes of old blocks and verify the integrity relative to other peers, see the ledgerutil verify and compare commands:
https://hyperledger-fabric.readthedocs.io/en/latest/commands/ledgerutil.html

The important aspect is that an administrator cannot modify their peer data and successfully pass it off to other nodes. The other nodes would detect the compromised data, for example if the modified peer tried to endorse a subsequent transaction the other nodes would detect the endorsement difference and invalidate the subsequent transaction.
